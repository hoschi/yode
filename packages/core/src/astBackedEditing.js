import profiler from './profiler'
import logger from './logger'
import * as R from 'ramda'
import parser from 'ast/parser-recast-jsx'
import { getFunctionIndexByText } from 'ast/compareFunctions'
import {getNextId}from'./customIdGenerator'

let {estraverse} = parser
function parse (...args) {
    //let stop = profiler.start('parse', true);
    let r = parser.parse.apply(parser, args)
    //stop();
    return r
}

export function printAst (...args) {
    //let stop = profiler.start('--- printAst', true);
    let r = parser.printAst.apply(parser, args)
    //stop();
    return r
}

export let getMetaData = (node) => ({
    ...R.pick(['hasConnectedError', 'syntaxError'], node),
    title: 'node-' + node.customId
})

export function addTextToNode (ast) {
    ast.text = printAst(ast)
}

export function isNodeDirty (node) {
    return node.text !== node.unformattedText
}

export function getNodeForFirstFoundType (type, ast) {
    let found
    estraverse.traverse(ast, {
        // search for node with same type as old function and save that one instead
        enter(node) {
            if (node.type === type) {
                found = node
                this.break()
            }
        }
    })
    return found
}

export function replaceNodeInAst (replacementNode, ast) {
    let modifiedAst = estraverse.replace(ast, {
        enter(node) {
            if (node.customId === replacementNode.customId) {
                return replacementNode
            }
        }
    })
    return modifiedAst
}

export function getAllContainerNodesRecursive (node, collected = []) {
    if (!node || node.isRoot) {
        return collected
    }
    return getAllContainerNodesRecursive(node.parentFunction, collected.concat([node]))
}

export function getAllChildrenNodesRecursive (nodes, collected = []) {
    if (R.isEmpty(nodes)) {
        return collected
    }

    return R.concat(
        nodes,
        R.chain((node) => {
            return getAllChildrenNodesRecursive(node.children, nodes.children)
        }, nodes)
    )
}

let cursorIsInNode = (cursor, node) => cursor > node.start && cursor < node.end

export function parseCode (text) {
    let stop = profiler.start('-- code to ast')
    let ast
    try {
        ast = parse(text)
    } /* eslint-disable */ catch ( error ) /* eslint-enable */ {
        logger.log(error)
        stop()
        return {
            error
        }
    }
    stop()
    return {
        ast
    }
}

export function getFunctionsFromAst (ast, fileId, functionsToCompare) {
    let stop = profiler.start('-- ast to functions')
    let functions = []
    let functionsToCompareLeft

    if (functionsToCompare) {
        functionsToCompareLeft = [].concat(functionsToCompare)
    }

    let parentFunc
    let currentFunc = {
        children: [],
        isRoot: true
    }
    let functionsTreeRoot = currentFunc

    let addNodeToParent = (node) => {
        parentFunc = currentFunc
        parentFunc.children.push(node)

        currentFunc = node
        // set empty children, in case this node has children from previous runs, which get now collected again
        currentFunc.children = []
        node.parentFunction = parentFunc
    }

    let leaveFunctionNode = () => {
        currentFunc = parentFunc
        parentFunc = currentFunc.parentFunction
    }

    estraverse.traverse(ast, {
        enter(node) {
            // code generated from FunctionExpression are not parseable again, skip for now
            if (node.type === 'FunctionDeclaration' || node.type === 'ArrowFunctionExpression') {
                addNodeToParent(node)
                if (node.unformattedText === node.text) {
                    // update both, because this can't disturb user input.
                    // Editor text and generated text are in sync. This is
                    // important when node.text changed from outside. E.g. when
                    // changing an inner function of this function node. In
                    // this case the generated text is not the text which is in
                    // the editor, because the ast changed, not the editor.
                    //
                    // TODO this can be removed when we track function containment, then we can replace the changed function AST in all containers not only file AST
                    addTextToNode(node)
                    node.unformattedText = node.text
                } else {
                    // always regenerate text for node, because this node could
                    // contain a function which was changed.
                    //
                    // TODO Could be skipped when `node.text` is already there
                    // and we can make sure this node contains no other
                    // function.
                    addTextToNode(node)
                }

                // AST nodes can't have syntax error, only unformatted code can
                node.syntaxError = undefined

                if (!node.unformattedText) {
                    // set unformatted (starting) text to node in case this is first run
                    node.unformattedText = node.text
                }

                // update or assign custom id
                if (node.customId) {
                    if (functionsToCompareLeft) {
                        // remove from compare list, we know this function by id
                        let currentFunctionIndex = functionsToCompareLeft.findIndex(f => f.customId === node.customId)
                        if (currentFunctionIndex >= 0) {
                            functionsToCompareLeft.splice(currentFunctionIndex, 1)
                        }
                    }
                } else {
                    // check if we previously known that function already
                    let foundFunctionIndex
                    if (functionsToCompareLeft) {
                        foundFunctionIndex = getFunctionIndexByText(node.text, functionsToCompareLeft)
                    }
                    if (foundFunctionIndex >= 0) {
                        let foundFunction = functionsToCompareLeft[foundFunctionIndex]
                        node.customId = foundFunction.customId
                        // restore dirty editor state if needed
                        if (isNodeDirty(foundFunction)) {
                            node.unformattedText = foundFunction.unformattedText
                            node.syntaxError = foundFunction.syntaxError
                        }
                        // remove function we found
                        functionsToCompareLeft.splice(foundFunctionIndex, 1)
                    } else {
                        node.customId = getNextId()
                    }
                }
                node.fileId = fileId
                // create new object to change identity, so redux subscriptions work as expected
                functions.push({
                    ...node
                })
            }
        },
        leave(node) {
            if (node.type === 'FunctionDeclaration' || node.type === 'ArrowFunctionExpression') {
                leaveFunctionNode(node)
            }
        }
    })

    // do some info logging
    if (functionsToCompareLeft && functionsToCompareLeft.length > 0) {
        logger.log('REMOVED FUNCTIONS', functionsToCompareLeft.length, functionsToCompareLeft)
        functionsToCompareLeft.forEach(function (node) {
            logger.log('++', node.text)
        })
    }

    if (functionsToCompareLeft && functionsToCompareLeft.length <= 0) {
        functionsToCompareLeft = undefined
    }

    stop()
    return {
        functions,
        functionsTreeRoot,
        removedFunctions:functionsToCompareLeft
    }
}

export function getInnerMostFunctionNode (sourceNode, cursor) {
    let ast,
        foundNode
    try {
        ast = parser.parse(sourceNode.unformattedText)
    } /* eslint-disable */ catch ( error ) /* eslint-enable */ {
        logger.log(error)
        return
    }
    parser.estraverse.traverse(ast, {
        leave(node) {
            // code generated from FunctionExpression are not parseable again, skip for now
            if (node.type === 'FunctionDeclaration' || node.type === 'ArrowFunctionExpression') {
                if (cursorIsInNode(cursor, node)) {
                    foundNode = node
                    this.break()
                }
            }
        }
    })
    return foundNode
}
