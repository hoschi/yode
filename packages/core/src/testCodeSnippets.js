export let letOnly = `
let only = 'simple'
`.trim()

export let nestedFunctions = `
let myVar = 'some string'
function myFunc (a) {
    function innerFunc () {
        return 'hello'
    }
    return a + innerFunc()
}
`.trim()

export let nestedFunctions_removedOneFunc = `
let myVar = 'some string'
function myFunc (a) {
    return a
}
`.trim()

export let jsxComponentAndTest = `
import React from 'react'

let HelloComponent = ({name}) => {
    return <span>Hello&nbsp;{name}!</span>;
}

describe('HelloComponent', () => {
    let name = 'Walter';

    it('displays name', () => {
        function getTestName(a, b) {
            return a + b + name;
        }

        let myName = getTestName('I', 'Am', name);
        expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
    })

    // function not recognized because not of a supported type
    it("doesn't work without name", function () {
        let myName = undefined;
        expect(<HelloComponent name={myName} />).contain('Hello !')
    });
})
`.trim()

export let jsxComponentAndTest_removedFuncs = `
import React from 'react'

let HelloComponent = ({name}) => {
    return <span>Hello&nbsp;{name}!</span>;
}
`.trim()

export let jsxClassesAndObjects = `
import React from 'react'

let getText = (name) => 'Hello ' + name;

class Welcome extends React.Component {
  // ClassMethod, not supported
  render() {
    let text = getText(this.props.name)
    return <h1>{text}</h1>;
  }
}

let MyObject = {
  name:'test',
  // ObjectMethod, not supported
  render() {
    let text = getText(this.name)
    return <span>{text}</span>;
  }
}
`.trim()
