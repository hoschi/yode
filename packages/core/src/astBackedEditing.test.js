let mockedId = 0
jest.mock('./customIdGenerator', () => {
    return {
        getNextId:() => ++mockedId
    }
})

import parser from 'ast/parser-recast-jsx'
import * as R from 'ramda'
import * as snip from './testCodeSnippets'
import * as s from './astBackedEditing'

beforeEach(() => {
    mockedId = 0
})

test('getFunctionsFromAst - without functions', function () {
    let ast = parser.parse(snip.letOnly)
    let result = s.getFunctionsFromAst(ast, 'myFile.js')

    expect(result.removedFunctions).toBeUndefined()
    expect(result.functions).toHaveLength(0)
    expect(result.functionsTreeRoot.children).toHaveLength(0)
    expect(result.functionsTreeRoot.isRoot).toBe(true)
})

test('getFunctionsFromAst - doesn\'t collect object functions', () => {
    let ast = parser.parse(snip.jsxClassesAndObjects)
    let result = s.getFunctionsFromAst(ast, 'myFile.js')

    expect(result.removedFunctions).toBeUndefined()
    expect(result.functions).toHaveLength(1)
    expect(result.functionsTreeRoot.children).toHaveLength(1)
    expect(result.functionsTreeRoot.isRoot).toBe(true)

    expect(result).toMatchSnapshot()
})

test('getFunctionsFromAst - collects functions flat and tree', () => {
    let ast = parser.parse(snip.jsxComponentAndTest)
    let result = s.getFunctionsFromAst(ast, 'myFile.js')

    expect(result.removedFunctions).toBeUndefined()
    expect(result.functions).toHaveLength(4)
    expect(result.functionsTreeRoot.children).toHaveLength(2)
    expect(result.functionsTreeRoot.isRoot).toBe(true)

    expect(result).toMatchSnapshot()
})

test('getFunctionsFromAst - removed one function on file level', () => {
    let ast, result
    ast = parser.parse(snip.nestedFunctions)
    result = s.getFunctionsFromAst(ast, 'myFile.js')
    expect(result.removedFunctions).toBeUndefined()
    expect(result.functions).toHaveLength(2)
    let oldFunctions = result.functions
    expect(result.functionsTreeRoot.children).toHaveLength(1)
    expect(result.functionsTreeRoot.isRoot).toBe(true)
    expect(result).toMatchSnapshot()

    ast = parser.parse(snip.nestedFunctions_removedOneFunc)
    result = s.getFunctionsFromAst(ast, 'myFile.js', oldFunctions)
    expect(result.removedFunctions).toHaveLength(1)
    expect(result.functions).toHaveLength(1)
    expect(result.functionsTreeRoot.children).toHaveLength(1)
    expect(result.functionsTreeRoot.isRoot).toBe(true)
    expect(result).toMatchSnapshot()
})

test('getFunctionsFromAst - remove more/nested functions on file level', () => {
    let ast, result

    ast = parser.parse(snip.jsxComponentAndTest)
    result = s.getFunctionsFromAst(ast, 'myFile.js')
    expect(result.removedFunctions).toBeUndefined()
    expect(result.functions).toHaveLength(4)
    let oldFunctions = result.functions
    expect(result.functionsTreeRoot.children).toHaveLength(2)
    expect(result.functionsTreeRoot.isRoot).toBe(true)
    expect(result).toMatchSnapshot()

    ast = parser.parse(snip.jsxComponentAndTest_removedFuncs)
    result = s.getFunctionsFromAst(ast, 'myFile.js', oldFunctions)
    expect(result.removedFunctions).toHaveLength(3)
    expect(result.functions).toHaveLength(1)
    expect(result.functionsTreeRoot.children).toHaveLength(1)
    expect(result.functionsTreeRoot.isRoot).toBe(true)
    expect(result).toMatchSnapshot()
})

test('getFunctionsFromAst - rerun keeps ids', () => {
    let simplifyNodes = R.map(R.pick(['customId', 'text']))
    let ast = parser.parse(snip.jsxComponentAndTest)
    let firstRun = s.getFunctionsFromAst(ast, 'myFile.js')

    expect(firstRun.removedFunctions).toBeUndefined()
    expect(firstRun.functions).toHaveLength(4)
    expect(firstRun.functionsTreeRoot.children).toHaveLength(2)
    expect(firstRun.functionsTreeRoot.isRoot).toBe(true)
    let firstRunFunctionsSimple = simplifyNodes(firstRun.functions)

    let additionalRun
    additionalRun = s.getFunctionsFromAst(ast, 'myFile.js', firstRun.functions)

    expect(additionalRun.removedFunctions).toBeUndefined()
    expect(additionalRun.functions).toHaveLength(4)
    expect(additionalRun.functionsTreeRoot.children).toHaveLength(2)
    expect(additionalRun.functionsTreeRoot.isRoot).toBe(true)
    expect(simplifyNodes(additionalRun.functions)).toEqual(firstRunFunctionsSimple)

    // no error when not all functions in ast are there to compare
    additionalRun = s.getFunctionsFromAst(ast, 'myFile.js', [firstRun.functions[0]])

    expect(additionalRun.removedFunctions).toBeUndefined()
    expect(additionalRun.functions).toHaveLength(4)
    expect(additionalRun.functionsTreeRoot.children).toHaveLength(2)
    expect(additionalRun.functionsTreeRoot.isRoot).toBe(true)
    expect(simplifyNodes(additionalRun.functions)).toEqual(firstRunFunctionsSimple)

    // even works without functions to compare, but can't track removed functions
    additionalRun = s.getFunctionsFromAst(ast, 'myFile.js')

    expect(additionalRun.removedFunctions).toBeUndefined()
    expect(additionalRun.functions).toHaveLength(4)
    expect(additionalRun.functionsTreeRoot.children).toHaveLength(2)
    expect(additionalRun.functionsTreeRoot.isRoot).toBe(true)
    expect(simplifyNodes(additionalRun.functions)).toEqual(firstRunFunctionsSimple)
})

