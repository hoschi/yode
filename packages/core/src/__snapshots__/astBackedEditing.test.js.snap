// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFunctionsFromAst - collects functions flat and tree 1`] = `
Object {
  "functions": Array [
    ArrowFunctionExpression 3:21-5:1 {
      "children": "[]",
      "customId": 1,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "({name}) => {
          return <span>Hello&nbsp;{name}!</span>;
      }",
    },
    ArrowFunctionExpression 7:27-24:1 {
      "children": "[
          {
            children: [
              {
                children: [],
                customId: 4,
              },
            ],
            customId: 3,
          },
        ]",
      "customId": 2,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "() => {
          let name = 'Walter';
      
          it('displays name', () => {
              function getTestName(a, b) {
                  return a + b + name;
              }
      
              let myName = getTestName('I', 'Am', name);
              expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
          })
      
          // function not recognized because not of a supported type
          it(\\"doesn't work without name\\", function () {
              let myName = undefined;
              expect(<HelloComponent name={myName} />).contain('Hello !')
          });
      }",
    },
    ArrowFunctionExpression 10:24-17:5 {
      "children": "[
          {
            children: [],
            customId: 4,
          },
        ]",
      "customId": 3,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "() => {
          function getTestName(a, b) {
              return a + b + name;
          }
      
          let myName = getTestName('I', 'Am', name);
          expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
      }",
    },
    FunctionDeclaration 11:8-13:9 {
      "children": "[]",
      "customId": 4,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "function getTestName(a, b) {
          return a + b + name;
      }",
    },
  ],
  "functionsTreeRoot(shortened)": Object {
    "children": "[
      {
        children: [],
        customId: 1,
      },
      {
        children: [
          {
            children: [
              {
                children: [],
                customId: 4,
              },
            ],
            customId: 3,
          },
        ],
        customId: 2,
      },
    ]",
    "isRoot": true,
  },
  "removedFunctions": undefined,
}
`;

exports[`getFunctionsFromAst - doesn't collect object functions 1`] = `
Object {
  "functions": Array [
    ArrowFunctionExpression 3:14-3:39 {
      "children": "[]",
      "customId": 1,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "name => 'Hello ' + name",
    },
  ],
  "functionsTreeRoot(shortened)": Object {
    "children": "[
      {
        children: [],
        customId: 1,
      },
    ]",
    "isRoot": true,
  },
  "removedFunctions": undefined,
}
`;

exports[`getFunctionsFromAst - remove more/nested functions on file level 1`] = `
Object {
  "functions": Array [
    ArrowFunctionExpression 3:21-5:1 {
      "children": "[]",
      "customId": 1,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "({name}) => {
          return <span>Hello&nbsp;{name}!</span>;
      }",
    },
    ArrowFunctionExpression 7:27-24:1 {
      "children": "[
          {
            children: [
              {
                children: [],
                customId: 4,
              },
            ],
            customId: 3,
          },
        ]",
      "customId": 2,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "() => {
          let name = 'Walter';
      
          it('displays name', () => {
              function getTestName(a, b) {
                  return a + b + name;
              }
      
              let myName = getTestName('I', 'Am', name);
              expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
          })
      
          // function not recognized because not of a supported type
          it(\\"doesn't work without name\\", function () {
              let myName = undefined;
              expect(<HelloComponent name={myName} />).contain('Hello !')
          });
      }",
    },
    ArrowFunctionExpression 10:24-17:5 {
      "children": "[
          {
            children: [],
            customId: 4,
          },
        ]",
      "customId": 3,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "() => {
          function getTestName(a, b) {
              return a + b + name;
          }
      
          let myName = getTestName('I', 'Am', name);
          expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
      }",
    },
    FunctionDeclaration 11:8-13:9 {
      "children": "[]",
      "customId": 4,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "function getTestName(a, b) {
          return a + b + name;
      }",
    },
  ],
  "functionsTreeRoot(shortened)": Object {
    "children": "[
      {
        children: [],
        customId: 1,
      },
      {
        children: [
          {
            children: [
              {
                children: [],
                customId: 4,
              },
            ],
            customId: 3,
          },
        ],
        customId: 2,
      },
    ]",
    "isRoot": true,
  },
  "removedFunctions": undefined,
}
`;

exports[`getFunctionsFromAst - remove more/nested functions on file level 2`] = `
Object {
  "functions": Array [
    ArrowFunctionExpression 3:21-5:1 {
      "children": "[]",
      "customId": 1,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "({name}) => {
          return <span>Hello&nbsp;{name}!</span>;
      }",
    },
  ],
  "functionsTreeRoot(shortened)": Object {
    "children": "[
      {
        children: [],
        customId: 1,
      },
    ]",
    "isRoot": true,
  },
  "removedFunctions": Array [
    ArrowFunctionExpression 7:27-24:1 {
      "children": "[
          {
            children: [
              {
                children: [],
                customId: 4,
              },
            ],
            customId: 3,
          },
        ]",
      "customId": 2,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "() => {
          let name = 'Walter';
      
          it('displays name', () => {
              function getTestName(a, b) {
                  return a + b + name;
              }
      
              let myName = getTestName('I', 'Am', name);
              expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
          })
      
          // function not recognized because not of a supported type
          it(\\"doesn't work without name\\", function () {
              let myName = undefined;
              expect(<HelloComponent name={myName} />).contain('Hello !')
          });
      }",
    },
    ArrowFunctionExpression 10:24-17:5 {
      "children": "[
          {
            children: [],
            customId: 4,
          },
        ]",
      "customId": 3,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "() => {
          function getTestName(a, b) {
              return a + b + name;
          }
      
          let myName = getTestName('I', 'Am', name);
          expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
      }",
    },
    FunctionDeclaration 11:8-13:9 {
      "children": "[]",
      "customId": 4,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "function getTestName(a, b) {
          return a + b + name;
      }",
    },
  ],
}
`;

exports[`getFunctionsFromAst - removed one function on file level 1`] = `
Object {
  "functions": Array [
    FunctionDeclaration 2:0-7:1 {
      "children": "[
          {
            children: [],
            customId: 2,
          },
        ]",
      "customId": 1,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "function myFunc(a) {
          function innerFunc () {
              return 'hello'
          }
          return a + innerFunc()
      }",
    },
    FunctionDeclaration 3:4-5:5 {
      "children": "[]",
      "customId": 2,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "function innerFunc() {
          return 'hello'
      }",
    },
  ],
  "functionsTreeRoot(shortened)": Object {
    "children": "[
      {
        children: [
          {
            children: [],
            customId: 2,
          },
        ],
        customId: 1,
      },
    ]",
    "isRoot": true,
  },
  "removedFunctions": undefined,
}
`;

exports[`getFunctionsFromAst - removed one function on file level 2`] = `
Object {
  "functions": Array [
    FunctionDeclaration 2:0-4:1 {
      "children": "[]",
      "customId": 2,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "function myFunc(a) {
          return a
      }",
    },
  ],
  "functionsTreeRoot(shortened)": Object {
    "children": "[
      {
        children: [],
        customId: 2,
      },
    ]",
    "isRoot": true,
  },
  "removedFunctions": Array [
    FunctionDeclaration 2:0-7:1 {
      "children": "[
          {
            children: [],
            customId: 2,
          },
        ]",
      "customId": 1,
      "fileId": "myFile.js",
      "syntaxError": undefined,
      "text": "function myFunc(a) {
          function innerFunc () {
              return 'hello'
          }
          return a + innerFunc()
      }",
    },
  ],
}
`;
